File: src/main/java/com/hoi4utils/ui/parser/ParserViewerController.java
package com.hoi4utils.ui.parser;


import com.hoi4utils.clausewitz.HOIIVFiles;
import com.hoi4utils.HOIIVUtils;
import com.hoi4utils.hoi4.country.Country;
import com.hoi4utils.hoi4.country.CountryTag$;
import com.hoi4utils.hoi4.focus.FocusTree;
import com.hoi4utils.script.AbstractPDX;
import com.hoi4utils.script.PDXScript;
import map.StrategicRegion;
import map.ResourcesFile;
import map.State;
import com.hoi4utils.parser.Node;
import com.hoi4utils.parser.Parser;
import com.hoi4utils.parser.ParserException;
import com.hoi4utils.ui.HOIIVUtilsAbstractController;
import com.hoi4utils.ui.JavaFXUIManager;
import com.hoi4utils.ui.pdxscript.PDXTreeViewFactory;
import com.hoi4utils.ui.pdxscript.StratRegionPDXEditorController;
import javafx.collections.FXCollections;
import javafx.fxml.FXML;
import javafx.scene.control.*;
import javafx.scene.layout.AnchorPane;
import scala.jdk.javaapi.CollectionConverters;

import javax.swing.*;
import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class ParserViewerController extends HOIIVUtilsAbstractController {

	@FXML
	public Label idVersion;
	@FXML
	public TextField parsePDXFileTextField;
	@FXML
	public Button browseButton;
	@FXML
	public Label pdxIdentifierLabel;
	@FXML
	public AnchorPane pdxTreeViewPane;
	@FXML
	private TextField searchTextField;  // user enters search text here
	// Suppose we have a ListView
	@FXML
	private ListView<PDXScript<?>> filesListView;
	@FXML
	private TextArea pdxNodeTextArea;
	@FXML
	private MenuItem saveMenuItem;
	
	private final List<PDXScript<?>> pdxScripts = new ArrayList<>();
	
	public ParserViewerController() {
		setFxmlResource("ParserViewer.fxml");
		setTitle("HOIIVUtils Parser Viewer");
	}

	@FXML
	void initialize() {
		includeVersion();

		// Then on list selection:
		filesListView.getSelectionModel().selectedItemProperty().addListener((obs, oldVal, newVal) -> {
			if (newVal != null) {
				// Create a tree view for the newly selected script
				TreeView<PDXScript<?>> pdxTreeView = PDXTreeViewFactory.createPDXTreeView(newVal);
				pdxTreeViewPane.getChildren().removeIf(node -> node instanceof TreeView);
				pdxTreeViewPane.getChildren().add(pdxTreeView);

				AnchorPane.setTopAnchor(pdxTreeView, 25.0);
				AnchorPane.setBottomAnchor(pdxTreeView, 0.0);
				AnchorPane.setLeftAnchor(pdxTreeView, 0.0);
				AnchorPane.setRightAnchor(pdxTreeView, 0.0);

				searchTextField.setOnAction(event -> {
					String searchTerm = searchTextField.getText();
					PDXTreeViewFactory.searchAndSelect(pdxTreeView, searchTerm);
				});

				pdxTreeView.setOnMouseClicked(event -> {
					// todo improve visual?
					PDXScript<?> selectedPDX = pdxTreeView.getSelectionModel().getSelectedItem().getValue();
					if (selectedPDX != null) {
						pdxNodeTextArea.setText(selectedPDX.toScript());
					}
				});
			}
		});
	}

	private void includeVersion() {
		idVersion.setText(HOIIVUtils.get("version").toString());
	}

	@FXML
	private void handlePDXFileBrowseAction() {
		File initialDirectory = HOIIVFiles.Mod.folder;
		File selected = JavaFXUIManager.openChooser(browseButton, initialDirectory, true);

		System.out.println(selected);

		if (selected != null) {
			pdxScripts.clear();

			if (selected.isDirectory()) {
				List<File> allPDXFiles = CollectionConverters.asJava(PDXScript.allPDXFilesInDirectory(selected));

				for (File file : allPDXFiles) {
					try {
						Parser parser = new Parser(file);
						var rootNode = parser.parse();
						if (rootNode == null) continue;

						// Identify if it’s a strategic_region, state, focus_tree, etc.
//						var firstChild = rootNode.nodeValue().list().apply(0);
						var firstChild = rootNode.toList().apply(0); 
						String pdxIdentifier = firstChild.name();

						AbstractPDX<?> pdx = switch (pdxIdentifier) {
							case "focus_tree"       -> new FocusTree(file);
							case "state"            -> new State(false, file);
							case "strategic_region" -> new StrategicRegion(file);

							// ...
							default -> null;
						};

						if (pdx != null && !pdx.isUndefined()) {
							pdxScripts.add(pdx);
						}
					} catch (ParserException e) {
						// handle errors
					}
				}
			}

			else {
				parsePDXFileTextField.setText(selected.getAbsolutePath());
				// Create the parser from the selected file.
				Parser pdxParser = new Parser(selected);
				try {
					Node rootNode = pdxParser.parse();
					if (rootNode == null) {
						JOptionPane.showMessageDialog(null, "Error: Selected focus tree not found in loaded focus trees.");
						return;
					}

					// If the root node is not a parent (i.e. its raw value is not a list), set the label to "[empty]" and return.
					if (!rootNode.isParent()) {
						pdxIdentifierLabel.setText("[empty]");
						return;
					}

					// Get the child list from the root node’s raw value.
					scala.Option<scala.collection.mutable.ListBuffer<Node>> maybeList = rootNode.$list();
					if (maybeList.isDefined()) {
						scala.collection.mutable.ListBuffer<Node> childList = maybeList.get();
						Node childPDXNode = childList.apply(0); // Scala's apply(0) returns the first element.
						String pdxIdentifier = childPDXNode.name();
						if (childList.length() == 1) {
							pdxIdentifierLabel.setText(pdxIdentifier);
						} else {
							pdxIdentifierLabel.setText(selected.getName());
						}

						AbstractPDX<?> pdx = null;
						if (pdxIdentifier.equals("focus_tree")) {
							pdx = new FocusTree(selected);
						} else if (pdxIdentifier.equals("state")) {
							pdx = new State(false, selected);
						} else if (selected.getParent().endsWith("countries")
								&& selected.getParentFile().getParent().endsWith("history")) {
							pdx = new Country(selected, CountryTag$.MODULE$.get(selected.getName().substring(0, 3)));
						} else if (pdxIdentifier.equals("resources")) {
							pdx = new ResourcesFile(selected);
						} else if (pdxIdentifier.equals("strategic_region")) {
							pdx = new StrategicRegion(selected);
						}

						if (pdx == null || pdx.isUndefined()) return;
						this.pdxScripts.add(pdx);
					} else {
						pdxIdentifierLabel.setText("[empty]");
					}

//					// Build a TreeView out of the rootScript
//					TreeView<PDXScript<?>> pdxTreeView = PDXTreeViewFactory.createPDXTreeView(pdx);
//
//					pdxTreeViewPane.getChildren().removeIf(node -> node instanceof TreeView);
//					pdxTreeViewPane.getChildren().add(pdxTreeView);
//					AnchorPane.setTopAnchor(pdxTreeView, 25.0);
//					AnchorPane.setBottomAnchor(pdxTreeView, 0.0);
//					AnchorPane.setLeftAnchor(pdxTreeView, 0.0);
//					AnchorPane.setRightAnchor(pdxTreeView, 0.0);
//
//					searchTextField.setOnAction(event -> {
//						String searchTerm = searchTextField.getText();
//						PDXTreeViewFactory.searchAndSelect(pdxTreeView, searchTerm);
//					});
				} catch (ParserException e) {
					throw new RuntimeException(e);
				}
			}

			// After loading all PDX scripts:
			filesListView.setItems(FXCollections.observableList(pdxScripts));
			// show pdx's node in side view when pdx item is selected in list view
			filesListView.setOnMouseClicked(event -> {
				PDXScript<?> selectedPDX = filesListView.getSelectionModel().getSelectedItem();
				if (selectedPDX != null) {
					// todo improve visual?
					pdxNodeTextArea.setText(selectedPDX.toScript());
				}
			});
			// yes this is after because the last statement can set an empty list in view
			if (pdxScripts.isEmpty()) return;

			/* do stuff with the pdx scripts */


		} else {
			pdxIdentifierLabel.setText("[not found]");
		}
	}

	@FXML
	private void handleSaveAction() {
		savePDX(this.pdxScripts);
	}

	private void savePDX(List<PDXScript<?>> pdxScripts) {
		for (PDXScript<?> pdx : pdxScripts) {
			pdx.savePDX(new File("Parser Viewer PDXScripts"));
		}
	}

//	@FXML
//	private void handleOpenReplAction() {
//		// Launch an Ammonite REPL, passing in the pdxScripts list
//		new ReplUIController().open(pdxScripts); 
//	}

	@FXML
	private void handlePDXListViewEditAll() {
		switch(pdxScripts.getFirst()) {
			case StrategicRegion sr -> {
				// Open the strategic region editor
				StratRegionPDXEditorController stratRegionEditor = new StratRegionPDXEditorController();
				List<StrategicRegion> list = new ArrayList<>();
				// java workaround
				pdxScripts.forEach(pdx -> {
					if (pdx instanceof StrategicRegion) {
						list.add((StrategicRegion) pdx);
					}
				});
				stratRegionEditor.open(list);
			}
			default -> {
				JOptionPane.showMessageDialog(null, "No editor available for this PDX type.");
			}
		}
	}
}

File: src/main/scala/com/hoi4utils/ui/parser/ParserViewerController.scala
package com.hoi4utils.ui.parser

import com.hoi4utils.{HOIIVFiles, HOIIVUtils}
import com.hoi4utils.hoi4.country.{Country, CountryTag}
import com.hoi4utils.hoi4.focus.FocusTree
import com.hoi4utils.parser.{Node, Parser, ParserException}
import com.hoi4utils.script.{AbstractPDX, PDXScript}
import com.hoi4utils.ui.{HOIIVUtilsAbstractController, JavaFXUIManager}
import com.hoi4utils.ui.pdxscript.{PDXTreeViewFactory, StratRegionPDXEditorController}
import com.map.{ResourcesFile, State, StrategicRegion}
import javafx.collections.FXCollections
import javafx.fxml.FXML
import javafx.scene.control.*
import javafx.scene.layout.AnchorPane

import javax.swing.JOptionPane
import java.io.File
import scala.collection.mutable.ListBuffer
import scala.compiletime.uninitialized
import scala.jdk.CollectionConverters.*
import scala.util.{Failure, Success, Try}

class ParserViewerController extends HOIIVUtilsAbstractController:

  @FXML var idVersion: Label = uninitialized
  @FXML var parsePDXFileTextField: TextField = uninitialized
  @FXML var browseButton: Button = uninitialized
  @FXML var pdxIdentifierLabel: Label = uninitialized
  @FXML var pdxTreeViewPane: AnchorPane = uninitialized
  @FXML private var searchTextField: TextField = uninitialized // user enters search text here
  @FXML private var filesListView: ListView[PDXScript[?]] = uninitialized
  @FXML private var pdxNodeTextArea: TextArea = uninitialized
  @FXML private var saveMenuItem: MenuItem = uninitialized

  private val pdxScripts: ListBuffer[PDXScript[?]] = ListBuffer.empty

  setFxmlResource("ParserViewer.fxml")
  setTitle("HOIIVUtils Parser Viewer")

  @FXML
  def initialize(): Unit =
    includeVersion()

    // Handle list selection
    filesListView.getSelectionModel.selectedItemProperty().addListener { (_, _, newVal) =>
      Option(newVal).foreach { selectedScript =>
        // Create a tree view for the newly selected script
        val pdxTreeView = PDXTreeViewFactory.createPDXTreeView(selectedScript)
        pdxTreeViewPane.getChildren.removeIf(_.isInstanceOf[TreeView[?]])
        pdxTreeViewPane.getChildren.add(pdxTreeView)

        AnchorPane.setTopAnchor(pdxTreeView, 25.0)
        AnchorPane.setBottomAnchor(pdxTreeView, 0.0)
        AnchorPane.setLeftAnchor(pdxTreeView, 0.0)
        AnchorPane.setRightAnchor(pdxTreeView, 0.0)

        searchTextField.setOnAction { _ =>
          val searchTerm = searchTextField.getText
          PDXTreeViewFactory.searchAndSelect(pdxTreeView, searchTerm)
        }

        pdxTreeView.setOnMouseClicked { _ =>
          Option(pdxTreeView.getSelectionModel.getSelectedItem).foreach { selectedItem =>
            val selectedPDX = selectedItem.getValue
            if selectedPDX != null then
              pdxNodeTextArea.setText(selectedPDX.toScript)
          }
        }
      }
    }

  private def includeVersion(): Unit =
    idVersion.setText(HOIIVUtils.get("version").toString)

  @FXML
  private def handlePDXFileBrowseAction(): Unit =
    val initialDirectory = HOIIVFiles.Mod.folder
    val selected = JavaFXUIManager.openChooser(browseButton, initialDirectory, true)

    println(selected)

    Option(selected) match
      case Some(file) =>
        pdxScripts.clear()

        if file.isDirectory then
          val allPDXFiles = PDXScript.allPDXFilesInDirectory(file).asJava.asScala.toList

          allPDXFiles.foreach { pdxFile =>
            Try {
              val parser = Parser(pdxFile)
              parser.parse
            } match
              case Success(rootNode) if rootNode != null =>
                val firstChild = rootNode.toList.head
                val pdxIdentifier = firstChild.name

                val pdx: Option[AbstractPDX[?]] = pdxIdentifier match
                  case "focus_tree"       => Some(FocusTree(pdxFile))
                  case "state"            => Some(State(false, pdxFile))
                  case "strategic_region" => Some(StrategicRegion(pdxFile))
                  case _                  => None

                pdx.foreach { p =>
                  if !p.isUndefined then
                    pdxScripts += p
                }
              case Failure(_: ParserException) =>
              // handle errors silently
              case _ =>
            // handle null rootNode or other cases
          }
        else
          parsePDXFileTextField.setText(file.getAbsolutePath)
          val pdxParser = Parser(file)

          Try(pdxParser.parse) match
            case Success(rootNode) if rootNode != null =>
              if !rootNode.isParent then
                pdxIdentifierLabel.setText("[empty]")
              else
                rootNode.$list() match
                  case Some(childList) if childList.nonEmpty =>
                    val childPDXNode = childList.apply(0)
                    val pdxIdentifier = childPDXNode.name

                    if childList.length == 1 then
                      pdxIdentifierLabel.setText(pdxIdentifier)
                    else
                      pdxIdentifierLabel.setText(file.getName)

                    val pdx: Option[AbstractPDX[?]] = pdxIdentifier match
                      case "focus_tree" =>
                        Some(FocusTree(file))
                      case "state" =>
                        Some(State(false, file))
                      case _ if file.getParent.endsWith("countries") &&
                        file.getParentFile.getParent.endsWith("history") =>
                        Some(Country(file, CountryTag.get(file.getName.substring(0, 3))))
                      case "resources" =>
                        Some(ResourcesFile(file))
                      case "strategic_region" =>
                        Some(StrategicRegion(file))
                      case _ =>
                        None

                    pdx.foreach { p =>
                      if p != null && !p.isUndefined then
                        pdxScripts += p
                    }
                  case _ =>
                    pdxIdentifierLabel.setText("[empty]")
            case Success(_) =>
              JOptionPane.showMessageDialog(null, "Error: Selected focus tree not found in loaded focus trees.")
            case Failure(e: ParserException) =>
              throw RuntimeException(e)
            case Failure(e) =>
              throw RuntimeException(e)

        // After loading all PDX scripts
        filesListView.setItems(FXCollections.observableList(pdxScripts.asJava))

        // Show pdx's node in side view when pdx item is selected in list view
        filesListView.setOnMouseClicked { _ =>
          Option(filesListView.getSelectionModel.getSelectedItem).foreach { selectedPDX =>
            pdxNodeTextArea.setText(selectedPDX.toScript)
          }
        }

        if pdxScripts.isEmpty then return

      /* do stuff with the pdx scripts */

      case None =>
        pdxIdentifierLabel.setText("[not found]")

  @FXML
  private def handleSaveAction(): Unit =
    savePDX(pdxScripts.toList)

  private def savePDX(pdxScripts: List[PDXScript[?]]): Unit =
    pdxScripts.foreach { pdx =>
      pdx.savePDX(File("Parser Viewer PDXScripts"))
    }

  // @FXML
  // private def handleOpenReplAction(): Unit =
  //   // Launch an Ammonite REPL, passing in the pdxScripts list
  //   ReplUIController().open(pdxScripts.toList)

  @FXML
  private def handlePDXListViewEditAll(): Unit =
    if pdxScripts.nonEmpty then
      pdxScripts.head match
        case _: StrategicRegion =>
          // Open the strategic region editor
          val stratRegionEditor = StratRegionPDXEditorController()
          val strategicRegions = pdxScripts.collect {
            case sr: StrategicRegion => sr
          }.toList
          stratRegionEditor.open(strategicRegions.asJava)
        case _ =>
          JOptionPane.showMessageDialog(null, "No editor available for this PDX type.")